lib/exabgp/application/healthcheck.py

```diff
@@ -472,7 +465,7 @@ def exabgp(target):
             setup_ips(options.ips, options.label, options.sudo)
 
         logger.info("send announces for %s state to ExaBGP", target)
-        metric = vars(options).get("{0}_metric".format(str(target).lower()))
+        metric = vars(options).get("{0}_metric".format(str(target).lower()), 0)
         for ip in options.ips:
             if options.withdraw_on_down:
```

Comment:
You need to use `options.withdraw_on_down or target == states.EXIT`.

Reply:
if withdraw_on_down is False, routes will not be withdrawed.

Reply:
That's why you need to *add* `or target == states.EXIT`. This way, the routes get withdrawn on exit whatever the value of `withdraw_on_down` is.

Reply:
I mean when user specify withdraw_on_down to False, user donot want route to be withdrawed under any condition.

Reply:
Maybe not on down, but on exit, yes. Isn't the feature you want to implement? `withdraw_on_down` is `False` by default.

Reply:
I just want route withdraw properly

Reply:
If I take your first commit and add this patch:

```diff
diff --git a/lib/exabgp/application/healthcheck.py b/lib/exabgp/application/healthcheck.py
index a11151c4887c..8631847646be 100644
--- a/lib/exabgp/application/healthcheck.py
+++ b/lib/exabgp/application/healthcheck.py
@@ -467,7 +467,7 @@ def loop(options):
         logger.info("send announces for %s state to ExaBGP", target)
         metric = vars(options).get("{0}_metric".format(str(target).lower()), 0)
         for ip in options.ips:
-            if options.withdraw_on_down:
+            if options.withdraw_on_down or target is states.EXIT:
                 command = "announce" if target is states.UP else "withdraw"
             else:
                 command = "announce"
```

I am able to get a withdraw on exit:

```
INFO[healthcheck] send announces for UP state to ExaBGP
announce route 203.0.113.14/32 next-hop self med 100
INFO[healthcheck] send announces for UP state to ExaBGP
announce route 203.0.113.14/32 next-hop self med 100
INFO[healthcheck] send announces for EXIT state to ExaBGP
withdraw route 203.0.113.14/32 next-hop self
```

Reply:
My test case：
1. initial config
```
process p1 {
	run python -m exabgp healthcheck  --ip IPADDR1 --withdraw-on-down --dynamic-ip-setup ;
	encoder text;
}

neighbor xxx {

	api {
		processes [ p1 ];
	}
}
```
2. change config to following：

```
process p2 {
	run python -m exabgp healthcheck  --ip IPADDR2 --withdraw-on-down --dynamic-ip-setup ;
	encoder text;
}

neighbor xxx {

	api {
		processes [ p2 ];
	}
}
```

3. reload config using exabgpcli
```
exabgpcli reload
```

Reply:
My test case：
1. start exabgp with initial config
```
process p1 {
	run python -m exabgp healthcheck  --ip IPADDR1 --withdraw-on-down --dynamic-ip-setup ;
	encoder text;
}

neighbor xxx {

	api {
		processes [ p1 ];
	}
}
```
2. change config to following：

```
process p2 {
	run python -m exabgp healthcheck  --ip IPADDR2 --withdraw-on-down --dynamic-ip-setup ;
	encoder text;
}

neighbor xxx {

	api {
		processes [ p2 ];
	}
}
```

3. reload config using exabgpcli
```
exabgpcli reload
```

Reply:
I've figure out that the cause is terminate process and bgp communication can not run parallelly. Terminate process need bgp communication to withdraw routes， but bgp communication code needs to wait termination done to continue。

Reply:
I've fix this using new thread to terminate process, but there maybe better ways to do this.

Reply:
Could you please explain why using thread is fixing the issue ?

Reply:
run termination  in a new thread make terminate_process call return immediately， thus the program can go up to process  bgp communication.